daml-foundations/daml-ghc/damldoc/src/DA/Daml/GHC/Damldoc/HaddockParse.hs

```diff
@@ -109,37 +116,61 @@ collectDocs = go Nothing []
     go Nothing docs (d:ds) = go (Just d) docs ds
     go (Just prev) docs (d:ds) = finished prev docs (go (Just d) [] ds)
 
-    finished decl docs rest = (unLoc decl, reverse docs) : rest
+    finished decl docs rest = (decl, toDocText . map unDocText . reverse $ docs) : rest
+
+-- | Context in which to extract a module's docs. This is created from
+-- 'TypecheckedModule' by 'buildDocCtx'.
+data DocCtx = DocCtx
+    { dc_mod :: Modulename
+    , dc_tcmod :: TypecheckedModule
+    , dc_decls :: [DeclData]
+    , dc_templates :: Set.Set Typename
```

Comment:
Might be nice to add newtypes around `Typename` for `TemplateName` and `ChoiceName`.